File: python/101-150/131. Palindrome Partitioning.py
class Solution(object):
    def partition(self, s):
        """
        :type s: str
        :rtype: List[List[str]]
        """
        res = []
        self.helper(s, [], res)
        return res
    def helper(self, s, temp, res):
        if not s:
            res.append(temp)
            return
        for i in range(1, len(s) + 1):
            if self.isPal(s[:i]):
                self.helper(s[i:], temp + [s[:i]], res)
    def isPal(self, s):
        return s == s[::-1]

File: cpp/101-150/131. Palindrome Partitioning.cpp
class Solution {
public:
    vector<vector<string>> partition(string s) {
        vector<vector<string>> res;
        if(s.empty())
            return res;
        vector<string> path;
        helper(0, res, s, path);
        return res;
    }
    bool isPanlindrom(string s, int i, int j) {
        while (i <= j) {
            if (s[i++] != s[j--])
                return false;
        }
        return true;
    }
    void helper(int index, vector<vector<string>> &res, string s, vector<string> &path) {
        if(index == s.size())
            res.push_back(path);
        for (int i = index; i < s.size(); i++) {
            if(isPanlindrom(s, index, i)){
                path.push_back(s.substr(index, i - index + 1));
                helper(i + 1, res, s, path);
                path.pop_back();
            }
        }
    }
};